parlai/agents/language_model/language_model.py

```diff
@@ -0,0 +1,480 @@
+        """Convert a list of observations into input & target tensors."""
+        labels = None
+        valid_inds = None
+        end_idxs = None
+        y_lens = None
+        if is_training:
+            for obs in observations:
+                if obs:
+                    if 'text2vec' in obs:
+                        self.next_batch += obs['text2vec']
+            if len(self.next_batch) <= (seq_len + 1) * self.batchsize:
+                return None, None, None, None, None, None
+            else:
+                data_list = []
+                targets_list = []
+                total = len(self.next_batch)//((seq_len + 1) * self.batchsize)
+                for _ in range(total):
+                    batch = self.next_batch[:((seq_len + 1) * self.batchsize)]
+                    self.next_batch = self.next_batch[((seq_len + 1) * self.batchsize):]
+
+                    source = torch.LongTensor(batch).view(self.batchsize, -1).t().contiguous()
+                    data = Variable(source[:seq_len])
+                    targets = Variable(source[1:].view(-1))
+
+                    if self.use_cuda:
+                        data = data.cuda()
+                        targets = targets.cuda()
+
+                    data_list.append(data)
+                    targets_list.append(targets)
+        else:
+            # here we get valid examples and pad them with zeros
+            xs, ys, labels, valid_inds, end_idxs, y_lens = PaddingUtils.pad_text(
+                observations, self.dict, self.END_IDX, self.NULL_IDX)
+            if self.use_cuda:
+                xs = Variable(xs).cuda()
+                ys = Variable(ys).cuda()
+            else:
+                xs = Variable(xs)
+                ys = Variable(ys)
+            data_list = [xs]
+            targets_list = [ys]
+
+        return data_list, targets_list, labels, valid_inds, end_idxs, y_lens
+
+    def batch_act(self, observations):
+        batch_reply = [{'id': self.getID()} for _ in range(len(observations))]
+        if 'labels' in observations[0]:
+            # if we are starting a new training epoch, reinitialize hidden
+            if self.is_training == False:
+                self.hidden = self.model.init_hidden(self.batchsize)
+            self.is_training = True
+            data_list, targets_list, _, _, end_idxs, y_lens = self.batchify(observations, self.opt['seq_len'], self.is_training)
+        else:
+            # if we just finished training, reinitialize hidden
+            if self.is_training == True:
+                self.hidden = self.model.init_hidden(self.batchsize)
+                self.is_training = False
+            data_list, targets_list, labels, valid_inds, end_idxs, y_lens = self.batchify(observations, self.opt['seq_len'], self.is_training)
+
+        if data_list is None:
+            # not enough data to batch act yet, return empty responses
+            return batch_reply
+
+        batch_reply = []
+        for i in range(len(data_list)):
```

Comment:
can you add comment along the lines of: `# data_list is a list of batches: always len 1 during eval, but could be zero or more than 1 during train`